lib/pure/includes/oserr.nim

```diff
@@ -3,14 +3,46 @@
 when not declared(os) and not declared(ospaths):
   {.error: "This is an include file for os.nim!".}
 
-when not defined(nimscript):
+when not weirdTarget:
   var errno {.importc, header: "<errno.h>".}: cint
 
-  proc c_strerror(errnum: cint): cstring {.
-    importc: "strerror", header: "<string.h>".}
-
   when defined(windows):
     import winlean
+  else:
+    # `strerror` is not necessarily thread-safe. Use `strerror_r` instead.
+    #
+    # glibc: https://linux.die.net/man/3/strerror_r
+    # The XSI-compliant version of strerror_r() is provided if:
+    # (_POSIX_C_SOURCE >= 200112L || _XOPEN_SOURCE >= 600) && ! _GNU_SOURCE
+    # Otherwise, the GNU-specific version is provided.
+    # Note: -D_GNU_SOURCE=0 also defines the GNU-specific version (docs issue)
+    {.emit: """
+    #include <errno.h>
+    #include <string.h>
+    #if !defined(__GLIBC__) || \
+        (((defined(_POSIX_C_SOURCE) && (_POSIX_C_SOURCE - 0) >= 200112L) || \
+        (defined(_XOPEN_SOURCE) && (_XOPEN_SOURCE - 0) >= 600)) && \
+        !defined(_GNU_SOURCE))
```

Comment:
In `nimbase.h` we have

```
#if defined(__GNUC__) && !defined(__ZEPHYR__)
/* Zephyr does some magic in it's headers that override the GCC stdlib. This breaks that. */
#  define _GNU_SOURCE 1
#endif
```

So we know _GNU_SOURCE is defined. This should simplify this logic quite a bit.

Reply:
Not sure that it can. `__GNUC__` is defined when compiling under GCC. `__GLIBC__` is defined when linking against glibc. Technically, could use GCC with other stdlibs as well (in which case `_GNU_SOURCE` does not have an effect). Also, even if `__GNUC__` is assumed to imply `__GLIBC__` there is still the `__ZEPHYR__` case which does not define `_GNU_SOURCE`.

Furthermore, deviating from the "official" feature test macro here (https://linux.die.net/man/3/strerror_r) means that there will be a dependency between `nimbase.h` and `oserr.nim` / if one of them changes the other one needs to at least be consulted for correctness.